"RuntimeError: dictionary changed size during iteration" on pypy

On pypy, `make test` fails in `NocaseDict.iterkeys()` with:
```
RuntimeError: dictionary changed size during iteration
```

This happens in 17 testcases, all with the same symptom.

It happens only on pypy, and there with minimum and latest package levels.
It happens only on the master branch, but not on the stable_0.17 branch.

Full error:

```
___ test_create_namespace_2[New top level namespace-kwargs0-None-None-True] ____
desc = 'New top level namespace'
kwargs = {'exp_namespace': 'abc', 'new_namespace': 'abc'}, exp_exc_types = None
exp_warn_types = None, condition = True
    def wrapper_func(desc, kwargs, exp_exc_types, exp_warn_types, condition):
        ...
>               test_func(testcase, **kwargs)  # not expecting an exception
tests/unittest/utils/pytest_extensions.py:165: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/unittest/pywbem/test_wbemserverclass.py:222: in test_create_namespace_2
    act_namespace = server.create_namespace(new_namespace)
pywbem/_server.py:503: in create_namespace
    self.conn.CreateInstance(ns_inst, namespace=self.interop_ns)
pywbem/_cim_operations.py:3152: in CreateInstance
    NewInstance=instance)
../../../virtualenv/pypy2.7-7.1.1/site-packages/mock/mock.py:1092: in __call__
    return _mock_self._mock_call(*args, **kwargs)
../../../virtualenv/pypy2.7-7.1.1/site-packages/mock/mock.py:1149: in _mock_call
    result = effect(*args, **kwargs)
pywbem_mock/_wbemconnection_mock.py:1125: in _mock_imethodcall
    result = methodnameattr(namespace, **params)
pywbem_mock/_wbemconnection_mock.py:1304: in _imeth_CreateInstance
    NewInstance=params['NewInstance'])
pywbem_mock/_defaultinstanceprovider.py:200: in CreateInstance
    namespace, NewInstance)
pywbem_mock/_defaultinstanceprovider.py:510: in CreateInstance
    validate_inst_props(namespace, target_class, new_instance)
pywbem_mock/_defaultinstanceprovider.py:114: in validate_inst_props
    for iprop_name in instance:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = NocaseDict({'SchemaUpdatesAllowed': CIMProperty(name='SchemaUpdatesAllowed', v...ue, overridable=True, translatable=True, toinstance=None, propagated=True)}))})
    def iterkeys(self):
        """
        Return an iterator through the dictionary keys in their original
        case, preserving the original order of items.
        """
>       for item in six.iteritems(self._data):
E       RuntimeError: dictionary changed size during iteration
pywbem/_nocasedict.py:331: RuntimeError
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -33,11 +33,11 @@ matrix:
       env:
         - PACKAGE_LEVEL=minimum
 
-    - os: linux
-      language: python
-      python: "2.7"
-      env:
-        - PACKAGE_LEVEL=latest
+#    - os: linux
+#      language: python
+#      python: "2.7"
+#      env:
+#        - PACKAGE_LEVEL=latest
 
 #    - os: linux
 #      language: python
@@ -105,11 +105,11 @@ matrix:
 #      env:
 #        - PACKAGE_LEVEL=minimum
 
-#    - os: linux
-#      language: python
-#      python: "pypy"  # currently Python 2.7.13, PyPy 7.1.1
-#      env:
-#        - PACKAGE_LEVEL=latest
+    - os: linux
+      language: python
+      python: "pypy"  # currently Python 2.7.13, PyPy 7.1.1
+      env:
+        - PACKAGE_LEVEL=latest
 
 # TODO: pypy3 disabled due to install error with typed-ast (from pylint)
 #    - os: linux

--- a/dev-requirements.txt
+++ b/dev-requirements.txt
@@ -54,7 +54,9 @@ lxml>=4.2.4; python_version >= '3.5' and python_version <= '3.7'
 lxml>=4.4.1; python_version >= '3.8'
 
 # Virtualenv
-virtualenv>=14.0.0; python_version < '3.5'
+# Virtualenv 20.0.19 does not install pip into the virtualenv on Python 3.4.
+# TODO: Remove the circumvention of pinning virtualenv once solved.
+virtualenv>=14.0.0,<20.0.19; python_version < '3.5'
 virtualenv>=16.1.0; python_version >= '3.5' and python_version < '3.8'
 virtualenv>=20.0.0; python_version >= '3.8'
 

--- a/dev-requirements.txt
+++ b/dev-requirements.txt
@@ -102,9 +102,9 @@ astroid>=2.3.3; python_version >= '3.5'
 # Python 3.4 support.
 # Requiring typed-ast>=1.4.0 for Python 3.8 since it addresses compile errors
 # with missing pgenheaders.h and duplicate definition of a struct.
-typed-ast>=1.3.0,<1.4.0; python_version == '3.4' and implementation_name == "cpython"
-typed-ast>=1.3.0; python_version >= '3.5' and python_version < '3.8' and implementation_name == "cpython"
-typed-ast>=1.4.0; python_version >= '3.8' and implementation_name == "cpython"
+typed-ast>=1.3.0,<1.4.0; python_version == '3.4'
+typed-ast>=1.3.0; python_version >= '3.5' and python_version < '3.8'
+typed-ast>=1.4.0; python_version >= '3.8'
 
 # Flake8 and dependents (no imports, invoked via flake8 script):
 flake8>=3.7.9

--- a/minimum-constraints.txt
+++ b/minimum-constraints.txt
@@ -162,8 +162,8 @@ pylint==2.4.4; python_version >= '3.5'
 astroid==1.4.9; python_version == '2.7'
 astroid==2.1.0; python_version == '3.4'
 astroid==2.3.3; python_version >= '3.5'
-typed-ast==1.3.0; python_version >= '3.4' and python_version < '3.8' and implementation_name == "cpython"
-typed-ast==1.4.0; python_version >= '3.8' and implementation_name == "cpython"
+typed-ast==1.3.0; python_version >= '3.4' and python_version < '3.8'
+typed-ast==1.4.0; python_version >= '3.8'
 
 # Flake8 and dependents (no imports, invoked via flake8 script):
 flake8==3.7.9

--- a/pywbem_mock/_defaultinstanceprovider.py
+++ b/pywbem_mock/_defaultinstanceprovider.py
@@ -138,10 +138,13 @@ def validate_inst_props(namespace, target_class, instance):
 
         # If property not in instance, add it from class and use default value
         # from class
+        add_props = dict()
         for cprop_name in target_class.properties:
             if cprop_name not in instance:
                 default_value = target_class.properties[cprop_name]
-                instance[cprop_name] = default_value
+                add_props[cprop_name] = default_value
+
+    instance.update(add_props)
 
 
 class ProviderDispatcher(BaseProvider):

--- a/tests/installtest/test_install.sh
+++ b/tests/installtest/test_install.sh
@@ -173,11 +173,9 @@ function make_virtualenv()
     pip list --format=columns 2>/dev/null || pip list 2>/dev/null
   fi
 
-  # If pip is reinstalled (=downgraded) before reinstalling setuptools and wheel,
-  # we get permission errors on Travis when running with minimum package levels.
+  run "pip install pip $PIP_OPTS" "Reinstalling pip with PACKAGE_LEVEL=$PACKAGE_LEVEL"
   run "pip install setuptools $PIP_OPTS" "Reinstalling setuptools with PACKAGE_LEVEL=$PACKAGE_LEVEL"
   run "pip install wheel $PIP_OPTS" "Reinstalling wheel with PACKAGE_LEVEL=$PACKAGE_LEVEL"
-  run "pip install pip $PIP_OPTS" "Reinstalling pip with PACKAGE_LEVEL=$PACKAGE_LEVEL"
 
   verbose "Virtualenv before actual install test:"
   verbose "Pip version: $(pip --version 2>&1)"

--- a/tests/installtest/test_install.sh
+++ b/tests/installtest/test_install.sh
@@ -28,7 +28,7 @@ function abspath()
     fi
 }
 
-DEBUG="false"
+DEBUG="true"
 VERBOSE="true"
 
 MYNAME=$(basename "$0")
@@ -155,7 +155,12 @@ function make_virtualenv()
     pip list --format=columns 2>/dev/null || pip list 2>/dev/null
   fi
 
-  run "virtualenv -p $python_cmd_path $envdir" "Creating virtualenv: $envdir"
+  if [[ "$DEBUG" == "true" ]]; then
+    virtualenv_debug_opts="-vvv"
+  else
+    virtualenv_debug_opts=""
+  fi
+  run "virtualenv -p $python_cmd_path $virtualenv_debug_opts $envdir" "Creating virtualenv: $envdir"
   run "source $envdir/bin/activate" "Activating virtualenv: $envdir"
 
   verbose "Virtualenv before reinstalling base packages:"

--- a/tests/installtest/test_install.sh
+++ b/tests/installtest/test_install.sh
@@ -136,25 +136,37 @@ function make_virtualenv()
   envnamep="${ENVPREFIX}$envname"
   envdir=$VIRTUALENV_DIR/$envnamep
 
-  echo "Saving current virtualenv: $VIRTUAL_ENV"
+  echo "Saving location of current virtualenv: $VIRTUAL_ENV"
   VIRTUAL_ENV_SAVED=$VIRTUAL_ENV
 
   if [[ -d $envdir ]]; then
+    verbose "Removing new virtualenv: $envdir"
     remove_virtualenv $envname
   fi
 
   python_cmd_path=$(which $PYTHON_CMD)
-  verbose "Creating virtualenv: $envdir"
-  verbose "..using this Python command: $python_cmd_path"
-  verbose "..that has this Python version: $($PYTHON_CMD --version 2>&1)"
-  run "virtualenv -p $python_cmd_path $envdir"
+  verbose "Before creating virtualenv: $envdir"
+  verbose "Python version: $($PYTHON_CMD --version 2>&1) from $(which $PYTHON_CMD)"
+  verbose "Pip version: $(pip --version 2>&1)"
+  if [[ "$DEBUG" == "true" ]]; then
+    echo "Debug: Python module path:"
+    python -c "import sys, pprint; pprint.pprint(sys.path)"
+    echo "Debug: Packages:"
+    pip list --format=columns 2>/dev/null || pip list 2>/dev/null
+  fi
 
+  run "virtualenv -p $python_cmd_path $envdir" "Creating virtualenv: $envdir"
   run "source $envdir/bin/activate" "Activating virtualenv: $envdir"
-  verbose "Python command in this virtualenv: $(which python)"
-  verbose "Python version in this virtualenv: $(python --version 2>&1)"
 
-  verbose "Packages in this virtualenv before reinstalling base packages:"
-  pip list --format=columns 2>/dev/null || pip list 2>/dev/null
+  verbose "Virtualenv before reinstalling base packages:"
+  verbose "Python version: $(python --version 2>&1) from $(which python)"
+  verbose "Pip version: $(pip --version 2>&1)"
+  if [[ "$DEBUG" == "true" ]]; then
+    echo "Debug: Python module path:"
+    python -c "import sys, pprint; pprint.pprint(sys.path)"
+    echo "Debug: Packages:"
+    pip list --format=columns 2>/dev/null || pip list 2>/dev/null
+  fi
 
   # If pip is reinstalled (=downgraded) before reinstalling setuptools and wheel,
   # we get permission errors on Travis when running with minimum package levels.
@@ -162,7 +174,9 @@ function make_virtualenv()
   run "pip install wheel $PIP_OPTS" "Reinstalling wheel with PACKAGE_LEVEL=$PACKAGE_LEVEL"
   run "pip install pip $PIP_OPTS" "Reinstalling pip with PACKAGE_LEVEL=$PACKAGE_LEVEL"
 
-  verbose "Packages in this virtualenv before actual install test:"
+  verbose "Virtualenv before actual install test:"
+  verbose "Pip version: $(pip --version 2>&1)"
+  verbose "Packages:"
   pip list --format=columns 2>/dev/null || pip list 2>/dev/null
 }
 
@@ -213,6 +227,7 @@ function run()
     echo "Debug: running in this shell: $cmd"
     eval "$cmd"
     rc=$?
+    echo "Debug: command returns: rc=$rc"
   else
     eval "$cmd" >$CMD_LOG_FILE 2>&1
     rc=$?
@@ -312,20 +327,15 @@ function assert_import_ok()
 {
   local module
   module="$1"
-  if [[ "$DEBUG" == "true" ]]; then
-    echo "Debug: module=$module"
-    details=""
-    #details="; sys.path=$(python -c 'import sys; print(sys.path)')"
-  else
-    details=""
-  fi
-  assert_run_ok "python -c \"import ${module}\"" "Python module '${module}' cannot be imported$details"
+  verbose "Checking for successful import of module: $module"
+  assert_run_ok "python -c \"import ${module}\"" "Python module '${module}' cannot be imported"
 }
 
 function assert_import_fails()
 {
   local module
   module="$1"
+  verbose "Checking for failing import of module: $module"
   assert_run_fails "python -c \"import ${module}\"" "Python module '${module}' can be imported but should have failed"
 }
 

